src/mscorlib/shared/System/Threading/ExecutionContext.cs

```diff
@@ -128,134 +112,239 @@ public static bool IsFlowSuppressed()
+                    OnValuesChanged(currentExecutionCtx1, previousExecutionCtx1);
+                }
             }
+            // Enregistrer edi to edi1
+            ExceptionDispatchInfo edi1 = edi;
+            // If exception was thrown by callback, rethrow it now original contexts are restored
+            edi1?.Throw();
         }
 
-        internal static void EstablishCopyOnWriteScope(Thread currentThread, ref ExecutionContextSwitcher ecsw)
+        internal static void OnValuesChanged(ExecutionContext previousExecutionCtx, ExecutionContext nextExecutionCtx)
         {
-            Debug.Assert(currentThread == Thread.CurrentThread);
+            Debug.Assert(previousExecutionCtx != nextExecutionCtx);
 
-            ecsw.m_ec = currentThread.ExecutionContext;
-            ecsw.m_sc = currentThread.SynchronizationContext;
-        }
+            // Collect Change Notifications 
+            IAsyncLocal[] previousChangeNotifications = previousExecutionCtx?.m_localChangeNotifications;
+            IAsyncLocal[] nextChangeNotifications = nextExecutionCtx?.m_localChangeNotifications;
 
-        private static void OnContextChanged(ExecutionContext previous, ExecutionContext current)
-        {
-            Debug.Assert(previous != null);
-            Debug.Assert(current != null);
-            Debug.Assert(previous != current);
+            // At least one side must have notifications
+            Debug.Assert(previousChangeNotifications != null || nextChangeNotifications != null);
 
-            foreach (IAsyncLocal local in previous.m_localChangeNotifications)
+            // Fire Change Notifications
+            try
             {
-                object previousValue;
-                object currentValue;
-                previous.m_localValues.TryGetValue(local, out previousValue);
-                current.m_localValues.TryGetValue(local, out currentValue);
+                if (previousChangeNotifications != null && nextChangeNotifications != null)
+                {
+                    // Notifications can't exist without values
+                    Debug.Assert(previousExecutionCtx.m_localValues != null);
+                    Debug.Assert(nextExecutionCtx.m_localValues != null);
+                    // Both contexts have change notifications, check previousExecutionCtx first
+                    foreach (IAsyncLocal local in previousChangeNotifications)
+                    {
+                        previousExecutionCtx.m_localValues.TryGetValue(local, out object previousValue);
+                        nextExecutionCtx.m_localValues.TryGetValue(local, out object currentValue);
 
-                if (previousValue != currentValue)
-                    local.OnValueChanged(previousValue, currentValue, true);
-            }
+                        if (previousValue != currentValue)
+                        {
+                            local.OnValueChanged(previousValue, currentValue, true);
```

Comment:
Nit: mind adding `contextChanged: ` to name the bool args in all of these cases?  I realize they weren't there before.

Reply:
Added